include/ofi_mr.h

```diff
@@ -141,4 +145,46 @@ int ofi_mr_map_verify(struct ofi_mr_map *map, uintptr_t *io_addr,
 		      size_t len, uint64_t key, uint64_t access,
 		      void **context);
 
+
+/*
+ * Memory registration cache
+ */
+
+struct ofi_mr_entry {
+	struct iovec		iov;
+	uint64_t		access;	/* TODO */
+	unsigned int		retired:1;
+	int			use_cnt;
+	struct dlist_entry	lru_entry;
+	uint8_t			data[];
+};
+
+struct ofi_mr_cache {
+	struct util_domain	*domain;
+	size_t			size;
+	size_t			entry_data_size;
+
+	RbtHandle		mr_tree;
+	struct dlist_entry	lru_list;
+
+	uint64_t		cached_cnt;
+	uint64_t		search_cnt;
+	uint64_t		delete_cnt;
+	uint64_t		hit_cnt;
+
+	int			(*add_region)(struct ofi_mr_cache *cache,
```

Comment:
Dmitry's changes add more fields to the mr_entry.  What data is missing?  The cache is intended to hide registrations from the user when they are not needed.  Regions that require explicit application registration (e.g. region registered as an RMA target) are not expected to go through the cache.

Because of how the merge feature works, there's not necessarily a direct relation between the input passed into search and the region that is added.